[H/3] Fix code passed into QuicConnection.CloseAsync and QuicStream.Abort

ManickaP opened the pull request:
Fixes the code passed into `QuicConnection.CloseAsync` and `QuicStream.Abort` which was by default -1. Quic supports only error codes up to 62 bits non-negative integer values: https://www.rfc-editor.org/rfc/rfc9000.html#integer-encoding.
Also added test which crashes the process (if running with DEBUG build of MsQuic) before this change and runs to completion afterwards.

Let me know if you want me to shuffle the code around, I'm not much familiar with the structure of your code base.

cc @amcasey 

Fixes #55196

Also this will be followed up in S.N.Quic with some checks (https://github.com/dotnet/runtime/issues/101385).

ManickaP committed:
Fix code passed into QuicConnection.CloseAsync and QuicStream.Abort

amcasey commented on src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs:
```diff
@@ -82,6 +82,11 @@ public override void Abort(ConnectionAbortedException abortReason)
+            // Allowed error codes are up to 62 bits non-negative integer values: https://www.rfc-editor.org/rfc/rfc9000.html#integer-encoding
+            if (resolvedErrorCode < 0 || resolvedErrorCode > ((1L << 62) - 1))
+            {
+                resolvedErrorCode = _context.Options.DefaultCloseErrorCode;
```
For other readers: unless overridden, this will be `H3_NO_ERROR = 0x0100`.

amcasey replied:
Moot - we no longer modify the specified value.

amcasey commented on src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3RequestTests.cs:
```diff
@@ -1606,6 +1606,63 @@ public async Task GET_ClientDisconnected_ConnectionAbortRaised()
 
+    [ConditionalFact]
+    [MsQuicSupported]
+    public async Task GET_ClientIdlesOut_ConnectionAbortRaised()
```
I'm assuming this used to trigger the assert described in https://github.com/dotnet/aspnetcore/issues/55196 and that's why it's a useful regression test.  As a standalone test, however, it might be useful to check the error code as well.  It could even set `DefaultCloseErrorCode` and/or `DefaultCloseErrorCode` to confirm that the configuration is respected.

amcasey replied:
When I try to run this locally, I get 
```
| System.ObjectDisposedException: Cannot access a disposed object.
| Object name: 'System.Net.Quic.QuicListener'.
|    at System.Net.Quic.QuicListener.DisposeAsync()
```

amcasey replied:
It also fails without the product change, so it's not related to the PR.

amcasey replied:
I've pushed a change that I think will do this but, as noted above, I can't fully validate locally.  Let's see what CI says.

amcasey reviewed (approved):
I think it would be nice to flesh out the test a little, but I'm okay merging as-is if this unblocks you.

amcasey committed:
Validate that configured error code is used

amcasey commented:
My new check is failing CI.  The test fails for me locally, but not until after the log message check succeeds, so I'm not sure why that's happening.  I'll investigate.

amcasey committed:
Reorder checks to mitigate timing issues

amcasey committed:
Change log validation method

amcasey commented:
The windows failure is known CI flakiness

amcasey commented:
Okay, looking more closely, the failure I'm seeing here is the same one I'm seeing locally and isn't related to the logging check.  I think the test is flaky in that it won't reliably cause the connection to be closed on the server side.

halter73 commented on src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3RequestTests.cs:
```diff
@@ -1606,6 +1605,90 @@ public async Task GET_ClientDisconnected_ConnectionAbortRaised()
+        {
+            await host.StartAsync();
+
+            var client = HttpHelpers.CreateClient(idleTimeout: TimeSpan.FromSeconds(1));
```
If you just want to test that `MultiplexedConnectionContext.ConnectionClosed` fires and that there's a log from the Quic transport with the expected error code, you could use `QuicConnection.ConnectAsync` and call `CloseAsync(errorCode)` rather than rely on `QuicConnectionOptions.IdleTimeout` like SocketsHttpHandler does. This is a little less end-to-end since it doesn't test the interaction with HttpClient, but it should help the test run faster and potentially make it more reliable.

I'm a little surprised that `await connectionClosedTcs.Task.DefaultTimeout();` is timing out. The default timeout is 30 seconds, right? Does msquic check for idle connections super infrequently? @nibanks

amcasey replied:
I ended up deleting the test since (a) we can't reliably and performantly test client-idles-out scenarios and (b) we no longer munge the error code.

halter73 commented on src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs:
```diff
@@ -502,6 +502,11 @@ public override void Abort(ConnectionAbortedException abortReason)
+        // Allowed error codes are up to 62 bits non-negative integer values: https://www.rfc-editor.org/rfc/rfc9000.html#integer-encoding
+        if (resolvedErrorCode < 0 || resolvedErrorCode > ((1L << 62) - 1))
+        {
+            resolvedErrorCode = _context.Options.DefaultStreamErrorCode;
```
Can we throw or at least log an error if an application explicitly tries to set an invalid error code using something like `HttpContext.Features.Get<IHttpResetFeature>().Reset(-1)`?

It should be possible to determine if the `resolvedErrorCode` is out of range because it was explicitly set to be out of range vs it simply not being configured.

amcasey replied:
Related: we use -1 as a default [here](https://github.com/dotnet/aspnetcore/blob/7412c71cf881392233b5652310ab2b9e7fdf71eb/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.FeatureCollection.cs#L18).

ManickaP replied:
> It should be possible to determine if the resolvedErrorCode is out of range because it was explicitly set to be out of range vs it simply not being configured.

I think that the implementation of `IProtocolErrorCodeFeature.Error` setter for Quic should throw `ArgumentOutOfRangeException`.

amcasey replied:
I didn't update `IHttpResetFeature` directly - it immediately aborts, which sets `IProtocolErrorCodeFeature.Error`, which is what actually causes problems.  But I can add a redundant check upstream if we think it's important.

amcasey replied:
I moved up the check in `Abort` to make it effectively the same as checking in `Reset`.

amcasey committed:
Throw rather than silently adjusting invalid error codes

I worked backwards from calls to `QuicConnection.CloseAsync` and `QuicStream.Abort`.

TODO: Update tests

amcasey committed:
Move validation earlier to make inconsistent states less likely

amcasey committed:
Add test coverage for IHttpResetFeature

amcasey committed:
Add test coverage for IProtocolErrorCodeFeature

amcasey committed:
Add test coverage for IStreamAbortFeature

amcasey committed:
Remove flaky test

amcasey commented:
I believe the PR is ready to go.

amcasey commented on src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3RequestTests.cs:
```diff
@@ -2031,6 +2030,33 @@ public async Task GET_GracefulServerShutdown_RequestCompleteSuccessfullyInsideHo
         }
     }
 
+    [Fact]
```
```suggestion
    [ConditionalFact]
    [MsQuicSupported]
```

amcasey committed:
Make ServerReset_InvalidErrorCode conditional on msquic

amcasey commented:
Ah, it was auto-merge from the previous design and my sign-off counts because it's technically Mana's PR.  Well, shout if you don't like the exceptions @halter73.
